### PR TITLE
Table function now supports coercion

### PR DESCRIPTION
**Pull request by anderson-joyle**
Issue #1204.
Re-pushing due to merge/rebase conflicts. https://github.com/microsoft/Power-Fx/pull/1282

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by CarlosFigueiraMSFT** [comment]
>                     else if (DType.TryUnionWithCoerce(exprType, childType, out var returnType, out var needCoercion))

Should this change be gated based on the new PFxV1Compat flag? Today this is valid: `AddColumns( [{a:1,b:2},{a:true,b:"2",c:3}], "c", b+100 )`; if the type of the table node in the expression starts including "c", then this expression won't be valid anymore.

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs:5229 in 54e5030. [](commit_id = 54e50307ae799439b6189995c0b386dab2e47701, deletion_comment = False)

**Comment by CarlosFigueiraMSFT** [comment]
>         /// <param name="coercionNeeded">True if union with coercion was called.</param>

successful

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Types/DType.cs:3603 in 54e5030. [](commit_id = 54e50307ae799439b6189995c0b386dab2e47701, deletion_comment = False)

**Comment by CarlosFigueiraMSFT** [comment]
>         public static bool TryUnionWithCoerce(DType argType1, DType argType2, out DType returnType, out bool coercionNeeded)

Does this need to be public?

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Types/DType.cs:3605 in 54e5030. [](commit_id = 54e50307ae799439b6189995c0b386dab2e47701, deletion_comment = False)

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ArgCoercion.txt [hunk 1]
```diff
@@ -55,4 +55,4 @@ Error({Kind:ErrorKind.InvalidArgument})
 3
 
 >> Last([ {a:2}, {a:true, b:true, c:20}, {a:2, b:2, c:true} ]).c
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ArgCoercion.txt:57]: >Last([ {a:2}, {a:true, b:true, c:20}, {a:2, b:2, c:true} ]).c [](http://example.com/codeflow?start=3&length=61)

This is a breaking change, so this should only be done with the new PowerFxV1CompatibilityRules feature enabled.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by anderson-joyle** [comment]
>         public static bool TryUnionWithCoerce(DType argType1, DType argType2, out DType returnType, out bool coercionNeeded)

Yes. This is called by Table and Binder.

---
In reply to: [1496350985](https://github.com/microsoft/Power-Fx/pull/1309#issuecomment-1496350985) [](http://example.com/codeflow?ancestors=1496350985,1496350985)

---
Refers to: src/libraries/Microsoft.PowerFx.Core/Types/DType.cs:3605 in 7233ea1. [](commit_id = 7233ea1c37dec7ab6138ca573d9b7d1b38a8a30a, deletion_comment = False)

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**